### PR TITLE
UDTF Column argument support

### PR DESCRIPTION
**Pull request by pearu**
This PR implements:
- The new UDTF API support
- Refactor Omnisci Array to Omnisci Buffer, Array, Column, OutputColumn.
- Various improvements to typesystem.

### CODE REVIEW

File: rbc/typesystem.py [hunk 1]
```diff
@@ -456,6 +475,32 @@ def is_complete(self):
...
+                if not m.is_concrete:
+                    return False
+        elif self.is_function:
+            if not self[0].is_concrete:
+                return False
+            for a in self[1]:
+                if not a.is_concrete:
+                    return False
+        elif self.is_void:
+            pass
```
Comment (guilhermeleobas) [inline rbc/typesystem.py:499]: Shouldn't this return something?

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    return sig, codegen
+
+
+@extending.intrinsic
+def omnisci_buffer_len_(typingctx, data):
+    sig = types.int64(data)
+
+    def codegen(context, builder, signature, args):
+        data, = args
+        assert data.opname == 'load'
```
Comment (guilhermeleobas) [inline rbc/omnisci_backend/omnisci_buffer.py:108]: Is this assertion required?

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+
+@extending.intrinsic
+def omnisci_buffer_getitem_(typingctx, data, index):
+    eltype = data.members[0].dtype
+    sig = eltype(data, index)
+
+    def codegen(context, builder, signature, args):
+        data, index = args
+        assert data.opname == 'load'
```
Comment (guilhermeleobas) [inline rbc/omnisci_backend/omnisci_buffer.py:147]: Same

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+@extending.intrinsic
+def omnisci_buffer_setitem_(typingctx, data, index, value):
+    sig = types.none(data, index, value)
+
+    def codegen(context, builder, signature, args):
+        zero = int32_t(0)
+
+        data, index, value = args
+
+        assert data.opname == 'load'
```
Comment (guilhermeleobas) [inline rbc/omnisci_backend/omnisci_buffer.py:193]: Same

File: rbc/omnisci_backend/omnisci_buffer.py (hunk 2 above)
Comment (guilhermeleobas) [inline rbc/omnisci_backend/omnisci_buffer.py:108]: Is this assertion required?

File: rbc/omnisci_backend/omnisci_buffer.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    def pointer(self):
+        ptr_type = self._params.get('pointer')
+        if ptr_type is not None:
+            return ptr_type
+        return typesystem.Type.pointer(self)
+
+
+class BufferType(types.Type):
+
```
[the remark this answers is not in the document]
Comment (pearu) [reply rbc/omnisci_backend/omnisci_buffer.py:56]: Fixed: added doc string.

File: rbc/typesystem.py (hunk 1 above)
Comment (pearu) [reply rbc/typesystem.py:499]: The last return statement returns True. The same applies for is_struct and is_function if blocks.

File: rbc/omnisci_backend/omnisci_buffer.py (hunk 2 above)
Comment (pearu) [reply rbc/omnisci_backend/omnisci_buffer.py:108]: I am uncertain. The subsequent `buf = data.operands[0]` relies on that `data.opname == 'load'`. If the opname is not `load`, using `operands[0]` will lead to unpredictable behaviour and it seems to me that tracking down this possible(?) scenario will be very difficult. So I think the assert statement can be removed if one can be certain that opname is `load`.

**Pull request merged**